src/abstractdataframe/abstractdataframe.jl

```diff
@@ -685,22 +685,32 @@ Base.vcat(df::AbstractDataFrame) = df
 Base.vcat(dfs::AbstractDataFrame...) = vcat(AbstractDataFrame[dfs...])
 
 Base.vcat(dfs::Vector{Void}) = dfs
+
 function Base.vcat{T<:AbstractDataFrame}(dfs::Vector{T})
     isempty(dfs) && return DataFrame()
-    coltyps, colnams, similars = _colinfo(dfs)
-
     res = DataFrame()
-    Nrow = sum(nrow, dfs)
-    for j in 1:length(colnams)
-        colnam = colnams[j]
-        col = similar(similars[j], coltyps[j], Nrow)
-
-        i = 1
-        for df in dfs
-            if haskey(df, colnam)
-                copy!(col, i, df[colnam])
+    nrows = sum(nrow, dfs)
+    for colnam in unique([(names(e) for e in dfs)...;])
+        k = Bool[haskey(e, colnam) for e in dfs]
+        c = vcat((e[colnam] for e in view(dfs, k))...)
+        if all(k)
+            col = c
+        else
+            col = if _isnullable(c)
+              similar(c, nrows)
```

Comment:
Ah, I misread the code. So indeed that's correct, just suboptimal. You could use `Base.return_types` as I described to avoid doing this when inference works, and only fall back to it when it fails.